### PR TITLE
Add ability to cancel the progress

### PR DESCRIPTION
**Pull request by boyvanamstel**
Brings over the cancel functionality added to Gifski recently.

cancel

I chose to copy over the `CustomButton` verbatim. Using it here might help fleshing out how it should behave as a standalone component later.

The cancel button currently only appears when a `Progress` instance has been set that is `isCancellable`.

### CODE REVIEW

**Comment by sindresorhus** [comment]
If I click the cancel button, I don't think we should reset the progress. but rather keep the progress. And maybe reduce the opacity of the label and progressbar to indicate it's disabled? Or maybe even better, reduce saturation?

**Comment by sindresorhus** [comment]
Unrelated to this PR, but I really wish we could do a better checkmark icon. The current one doesn't fit in at all with the nice and thin `X`.

**Comment by sindresorhus** [comment]
I think we should support the cancel button when `.progressValue` is used too. There are many cases where using the `Progress` API doesn't make sense. Then we need `.isCancellable` and `.onCancel` on the CircularProgress instance. If a `Progress` instance is specified, the `CircularProgress#isCancellable` takes priority over `progress.isCancellable`.

File: Sources/CircularProgress/CustomButton.swift [hunk 1]
```diff
@@ -0,0 +1,327 @@
+import Cocoa
+
+// TODO(sindresorhus): I plan to extract this into a reusable package when it's more mature.
+
+extension CALayer {
+	// TODO: Find a way to use a strongly-typed KeyPath here.
+	// TODO: Accept NSColor instead of CGColor.
```
Comment (sindresorhus) [inline Sources/CircularProgress/CustomButton.swift:7]: If you have any ideas how to solve these, I would appreciate the help. No worries if not though.

**Comment by boyvanamstel** [comment]
> Unrelated to this PR, but I really wish we could do a better checkmark icon. The current one doesn't fit in at all with the nice and thin `X`.

Yea, I know. It bothers me too. There might not be a way around adding a `successLayer`.

**Comment by boyvanamstel** [comment]
> I think we should support the cancel button when `.progressValue` is used too. There are many cases where using the `Progress` API doesn't make sense. Then we need `.isCancellable` and `.onCancel` on the CircularProgress instance. If a `Progress` instance is specified, the `CircularProgress#isCancellable` takes priority over `progress.isCancellable`.

Makes sense. I added `.onCancel` in one of the commits, so I can just bring that back. I kinda like how simple the API is by just leveraging `Progress`, but I trust you when you say there are use cases where `Progress` isn't the best option.

**Comment by sindresorhus** [comment]
> Yea, I know. It bothers me too. There might not be a way around adding a successLayer.

Moved the discussion to a separate issue: https://github.com/sindresorhus/CircularProgress/issues/5

**Comment by boyvanamstel** [comment]
I've added the option to cancel either manual or progress based views.

Instead of `.onCancel` I've added `.onCancelled` because it only fires when cancellation was successful.

Cancellation is disabled by default to not break existing projects.

I've made `.isCancelled` and `.isFinished` `@IBInspectible` so you can see what they look like. Currently `.isCancelled` doesn't change the appearance, I've implemented an opacity change in the example project through `.isCancelled`. We could provide a default state through desaturation or an opacity change later.

cancellable

File: Example/AppDelegate.swift [hunk 2]
```diff
@@ -23,12 +23,13 @@ final class AppDelegate: NSObject, NSApplicationDelegate {
 	}
 
 	private func configureManualView() {
-		manualCircularProgress.onCancelled = {
-			print("Manual cancelled!")
+		manualCircularProgress.onCancelled = { circularProgress in
+			circularProgress.alphaValue = 0.3
```
Comment (sindresorhus) [inline Example/AppDelegate.swift]: Why not just `manualCircularProgress.alphaValue = 0.3`? I usually pass on self in closures too, but is it really that beneficial?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: readme.md [hunk 3]
```diff
@@ -119,6 +121,31 @@ Reset the progress back to zero without animating.
...
 
+/**
+Cancels `Progress` if it's set and prevents further updates.
+*/
+func cancelProgress() {}
+
+/**
+Triggers when the progress was cancelled succesfully.
+*/
+var onCancelled: (() -> Void)?
```
Comment (sindresorhus) [inline readme.md:132]: This needs to be updated if we decide to keep the parameter: https://github.com/sindresorhus/CircularProgress/pull/4#discussion_r250964699

File: Example/AppDelegate.swift (hunk 2 above)
Comment (boyvanamstel) [reply Example/AppDelegate.swift]: I was juggling a bunch of circular progresses in the example app at some point. It cleaned up the code a bit then. I think it's better to not have it moving forward.

File: readme.md (hunk 3 above)
Comment (boyvanamstel) [reply readme.md:132]: Parameter has been removed.

File: Sources/CircularProgress/CircularProgress.swift [hunk 4]
```diff
@@ -61,22 +74,67 @@ public final class CircularProgress: NSView {
...
 				progressObserver = progressInstance.observe(\.fractionCompleted) { sender, _ in
+					guard !self.isCancelled && !sender.isFinished else {
+						return
+					}
+
 					self.progress = sender.fractionCompleted
 				}
+
+				finishedObserver = progressInstance.observe(\.isFinished) { sender, _ in
+					guard !sender.isCancelled && sender.isFinished else {
```
Comment (sindresorhus) [inline Sources/CircularProgress/CircularProgress.swift]:
```suggestion
					guard !self.isCancelled && sender.isFinished else {
```

For consistency?

**Comment by boyvanamstel** [comment]
Rebased on master.

This is a warning now btw. Do you know how to fix it, besides putting it back on a single line?

screenshot 2019-01-27 at 10 43 40

**Comment by sindresorhus** [comment]
> Do you know how to fix it, besides putting it back on a single line?

https://github.com/sindresorhus/CircularProgress/pull/4/commits/1e9bb8b2b59610bc8ce09d4d7913e75112945599

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
Yay! This is awesome. 🎉

**Comment by sindresorhus** [comment]
@boyvanamstel Do you want to bring these changes back to Gifski or should I?

**Comment by boyvanamstel** [comment]
> @boyvanamstel Do you want to bring these changes back to Gifski or should I?

I'll open a PR later today, or tomorrow. 👍

**Comment by sindresorhus** [comment]
I noticed a tiny issue with the cancel button. If I click and hold it, drag the mouse outside the button, and then release, it will stay filled (active state), and the next time I hover it, it will still be filled.

link style

**Comment by boyvanamstel** [comment]
Yea, that happens with the show in Finder button in Gifski as well. We'll need to fix that in `CustomButton`.

**Pull request merged**